braindecode/preprocessing/preprocess.py

```diff
@@ -120,54 +127,109 @@ def preprocess(concat_ds, preprocessors):
         A concat of BaseDataset or WindowsDataset datasets to be preprocessed.
     preprocessors: list(Preprocessor)
         List of Preprocessor objects to apply to the dataset.
+    save_dir : str | None
+        If a string, the preprocessed data will be saved under the specified
+        directory and the datasets in ``concat_ds`` will be reloaded with
+        `preload=False`.
+    overwrite : bool
+        When `save_dir` is provided, controls whether to overwrite existing
+        files in `save_dir`.
+    n_jobs : int | None
+        Number of jobs for parallel execution.
 
     Returns
     -------
     BaseConcatDataset:
         Preprocessed dataset.
     """
+    # In case of serialization, make sure directory is available before
+    # preprocessing
+    if save_dir is not None and not overwrite:
+        check_save_dir_empty(save_dir)
+
     if not isinstance(preprocessors, Iterable):
         raise ValueError(
             'preprocessors must be a list of Preprocessor objects.')
     for elem in preprocessors:
         assert hasattr(elem, 'apply'), (
             'Preprocessor object needs an `apply` method.')
 
-    # get the preprocessing keyword arguments
+    list_of_ds = Parallel(n_jobs=n_jobs)(
+        delayed(_preprocess)(ds, i, preprocessors, save_dir, overwrite)
+        for i, ds in enumerate(concat_ds.datasets))
+
+    if save_dir is not None:  # Reload datasets and replace in concat_ds
+        concat_ds_reloaded = load_concat_dataset(
+            save_dir, preload=False, target_name=None)
+        _replace_inplace(concat_ds, concat_ds_reloaded)
+    else:
+        if n_jobs is None or n_jobs == 1:  # joblib did not make copies, the
+            # preprocessing happened in-place
+            # Recompute cumulative sizes as transforms might have changed them
+            concat_ds.cumulative_sizes = concat_ds.cumsum(concat_ds.datasets)
+        else:  # joblib made copies
+            _replace_inplace(concat_ds, BaseConcatDataset(list_of_ds))
+
+    # Store preprocessing keyword arguments in the dataset
     preproc_kwargs = _get_preproc_kwargs(preprocessors)
+    concat_kind = 'raw' if hasattr(concat_ds.datasets[0], 'raw') else 'window'
+    # XXX The kwargs are not currently saved along the datasets when `save_dir`
```

Comment:
Why not? `BaseConcatDataset.save()` should take care of it, see https://github.com/braindecode/braindecode/blob/master/braindecode/datasets/base.py#L422-L427.

Reply:
It's because I hadn't implemented it yet and I didn't want to forget :P The latest commit should cover that.